src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/simple/RandomizedJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,104 @@
+
+import java.util.List;
+import java.util.Random;
+
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.utils.PhysicalPlanFactory;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+
+public class RandomizedJoinPlanOptimizerImpl extends JoinPlanOptimizerBase {
+	@Override // Implement initializeEnumerationAlgorithm from JoinPlanOptimizerBase.
+	public EnumerationAlgorithm initializeEnumerationAlgorithm( final List<PhysicalPlan> subplans ) {
+		return new RandomizedEnumerationAlgorithm(subplans);
+	}
+	
+	protected class RandomizedEnumerationAlgorithm implements EnumerationAlgorithm {
+		/**
+		 * Have a list of subplans.
+		 */
+		protected final List<PhysicalPlan> subplans;
+
+		/**
+		 *  When initializing the algorithm, start with the given list of subplans.
+		 *  @param subplans
+		 *  (These @ things where create automatically when I was making the comments, should I keep them in? // Simon)
+		 */
+		public RandomizedEnumerationAlgorithm( final List<PhysicalPlan> subplans ) {
+			this.subplans = subplans;
+		}
+		
+
+		/**
+		 * Also taken from Greedy and left unchanged save for the different method being called within.
+		 */
+		@Override
+		public PhysicalPlan getResultingPlan() throws QueryOptimizationException {
+			PhysicalPlan currentPlan = chooseFirstSubplan();
+
+			while ( subplans.size() > 0 ){
+				currentPlan = addNextRandomJoin(currentPlan);
+			}
+
+			return currentPlan;
+		}
+
+		
+		/**
+		 * Creating all possible plans, which are then to be picked from at random.
+		 * This method is directly copied from GreedyJoinPlanOptimizer without modification.
+		 * @param currentPlan
+		 * @return
+		 */
+		protected PhysicalPlan[] createNextPossiblePlans( final PhysicalPlan currentPlan ) {
+			final PhysicalPlan[] plans = new PhysicalPlan[ subplans.size() ];
+			for ( int i = 0; i < subplans.size(); ++i ) {
+				plans[i] = PhysicalPlanFactory.createPlanWithJoin( currentPlan, subplans.get(i) );
+			}
+			return plans;
+		}
+		
+
+		/**
+		 * Chooses the first subplan.
+		 * Instead of the best plan chosen by Greedy, this method picks a completely random plan.
+		 * @return
+		 * @throws QueryOptimizationException
+		 */
+		protected PhysicalPlan chooseFirstSubplan() throws QueryOptimizationException {
+			Random random = new Random(); // Create a random seed.
+			int indexOfRandomPlan = random.nextInt(subplans.size()); // nextInt returns an int between 0 (inclusive) and size (exclusive).
+
+			final PhysicalPlan randomPlan = subplans.get(indexOfRandomPlan);
+			subplans.remove(indexOfRandomPlan);
+			return randomPlan;
+		}
+		
+
+		/**
+		 * Next random join.
+		 * @param currentPlan
+		 * @return
+		 * @throws QueryOptimizationException
+		 */
+		protected PhysicalPlan addNextRandomJoin( final PhysicalPlan currentPlan )
+				throws QueryOptimizationException
+		{
+			final PhysicalPlan[] nextPossiblePlans = createNextPossiblePlans(currentPlan);

```

Comment:
For this randomized JoinPlanOptimizer, there is no need to create all next possible plans at each step. Instead, just randomly pick one of the remaining subplans and create the next join plan with this one only (in the way it is done within the loop in `createNextPossiblePlans`).